src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,139 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+            }
+
+            if (statement.GetLastToken() != embeddedStatement.GetLastToken())
+            {
+                if (statement is IfStatementSyntax ifStatement && ifStatement.Statement == embeddedStatement)
+                {
+                    // The embedded statement is followed by an 'else' clause, which may span multiple lines without
+                    // triggering a braces requirement, such as this:
+                    //
+                    //   if (true)
+                    //     return;
+                    //   else          // <-- this else clause is two lines, but is not considered a multiline context
+                    //     return;
+                    //
+                    // ---
+                    // INTENTIONAL FALLTHROUGH
+                }
+                else
+                {
+                    var firstTokenAfterEmbeddedStatement = embeddedStatement.GetLastToken().GetNextToken();
+                    if (!FormattingRangeHelper.AreTwoTokensOnSameLine(firstTokenAfterEmbeddedStatement, statement.GetLastToken()))
+                    {
+                        // The part of the statement following the embedded statement does not fit on one line. Examples:
+                        //
+                        //   do
+                        //     SomeMethod();
+                        //   while (x < 0 ||    // <-- This condition is split across multiple lines.
+                        //          x > 10);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool RequiresBracesToMatchContext(SyntaxNode statement, StatementSyntax embeddedStatement)
+        {
+            if (!statement.IsKind(SyntaxKind.IfStatement, SyntaxKind.ElseClause))
+            {
+                // 'if' statements are the only statements that can have multiple embedded statements which are
+                // considered relative to each other.
+                return false;
+            }
+
+            var topLevelContext = GetTopLevelContext(statement);
+            if (AnyPartUsesBraces(topLevelContext))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static IfStatementSyntax GetTopLevelContext(SyntaxNode statement)
+        {
+            IfStatementSyntax result;
+            if (statement.IsKind(SyntaxKind.ElseClause))
+            {
+                result = (IfStatementSyntax)statement.Parent;
+            }
+            else
+            {
+                Debug.Assert(statement.IsKind(SyntaxKind.IfStatement));
+                result = (IfStatementSyntax)statement;
+            }
+
+            while (result != null)
+            {
+                if (!result.IsParentKind(SyntaxKind.ElseClause))
+                {
+                    break;
+                }
+
+                result = (IfStatementSyntax)result.Parent.Parent;
```

Comment:
note: i'm a little wary of this due to the implicit dependency it takes on how the parser works.  But i don't feel strongly it needs to change.

Reply:
➡️ I'll leave this for now. If it ever does fail it would cause a gold bar, but I'm not aware of any open language change proposals that would impact this behavior in the foreseeable future.

Reply:
It's less about language changes, and more about parser recovery.  For example, parser *almost* just took a change that would allow things like a top-level else.  I changed it though so that if you do just have a top-level else, that it will parse out a missing 'if-header' portion.

--

BTW, that's a good test case :)  can you verify you don't crash if the user only has `else Whatever();` in their code?

Reply:
Added a test in f5521e62f40a453201f95fcc64ce8fff749ac974